File: C1.UWP/CS/BasicLibrarySamples/View/C1CollectionView/Data/BaseObject.cs
using System;
using System.Reflection;
using System.Collections.Generic;
using System.ComponentModel;

namespace BasicLibrarySamples
{
    /// <summary>
    /// Object that implements INotifyPropertyChanged.
    /// </summary>
    public class BaseObject : INotifyPropertyChanged
    {
        // ** INotifyPropertyChanged
        public event PropertyChangedEventHandler PropertyChanged;
        protected virtual void OnPropertyChanged(string propName)
        {
            if (PropertyChanged != null)
                PropertyChanged(this, new PropertyChangedEventArgs(propName));
        }
        protected bool SetValue<T>(ref T field, T value, string propertyName)
        {
            if (!EqualityComparer<T>.Default.Equals(field, value))
            {
                field = value;
                OnPropertyChanged(propertyName);
                return true;
            }
            return false;
        }
    }

    /// <summary>
    /// Object that implements INotifyPropertyChanged and IEditableObject.
    /// </summary>
    public class BaseEditableObject : BaseObject, IEditableObject
    {
        object _clone;
        public void BeginEdit()
        {
            _clone = this.MemberwiseClone();
        }
        public void CancelEdit()
        {
            if (_clone != null)
            {
                foreach (var pi in this.GetType().GetRuntimeProperties())
                {
                    if (pi.CanRead && pi.CanWrite)
                    {
                        var value = pi.GetValue(_clone);
                        pi.SetValue(this, value);
                    }
                }
            }
            _clone = null;
        }
        public void EndEdit()
        {
            _clone = null;
        }
    }

}

File: C1.UWP/VB/BasicLibrarySamples/View/C1CollectionView/Data/BaseObject.vb
Imports System.ComponentModel
Imports System.Collections.Generic
Imports System.Reflection
Imports System

''' <summary>
''' Object that implements INotifyPropertyChanged.
''' </summary>
Public Class BaseObject
    Implements INotifyPropertyChanged
    ' ** INotifyPropertyChanged
    Public Event PropertyChanged As PropertyChangedEventHandler Implements INotifyPropertyChanged.PropertyChanged

    Protected Overridable Sub OnPropertyChanged(propName As String)
        RaiseEvent PropertyChanged(Me, New PropertyChangedEventArgs(propName))
    End Sub
    Protected Function SetValue(Of T)(ByRef field As T, value As T, propertyName As String) As Boolean
        If Not EqualityComparer(Of T).[Default].Equals(field, value) Then
            field = value
            OnPropertyChanged(propertyName)
            Return True
        End If
        Return False
    End Function
End Class

''' <summary>
''' Object that implements INotifyPropertyChanged and IEditableObject.
''' </summary>
Public Class BaseEditableObject
    Inherits BaseObject
    Implements IEditableObject
    Private _clone As Object

    Public Sub BeginEdit() Implements IEditableObject.BeginEdit
        _clone = Me.MemberwiseClone()
    End Sub

    Public Sub CancelEdit() Implements IEditableObject.CancelEdit
        If _clone IsNot Nothing Then
            For Each pi As PropertyInfo In Me.[GetType]().GetRuntimeProperties()
                If pi.CanRead AndAlso pi.CanWrite Then
                    Dim value As Object = pi.GetValue(_clone)
                    pi.SetValue(Me, value)
                End If
            Next
        End If
        _clone = Nothing
    End Sub

    Public Sub EndEdit() Implements IEditableObject.EndEdit
        _clone = Nothing
    End Sub
End Class